design-proposals/cloud-hypervisor-integration.md

```diff
@@ -0,0 +1,291 @@
+- The two acceptable entries would be either `cloud-hypervisor` or `libvirt`.
+
+The `virt-operator` can inform all other components about the hypervisor type
+based on the information from the VMI spec. A different `virt-launcher` image
+would be picked instead of the default one, so that it contains `ch-launcher`.
+
+## Features
+
+### Supported features
+
+Here is a list of features expected to be available with Cloud Hypervisor.
+
+#### Lifecycle
+
+- Create a VM
+- Start a VM
+- Pause/Resume a VM
+- Snapshot/Restore a VM
+- Stop a VM
+
+#### Virtual Hardware
+
+- OVMF support (EFI)
+- CPU topology
+- CPU model is exclusively the equivalent of host for QEMU (no emulation of specific CPU model)
+- RNG using virtio-rng
+- Only headless VMs as we don't have graphics or video device emulation
+- CPU constraints + hotplug
+- Memory constraints + hotplug
+- Hugepages
+
+#### NUMA
+
+- Host NUMA to select specific host CPUs and make sure memory is allocated on expected NUMA node
+- Guest NUMA to expose any NUMA configuration to the guest
+
+#### Disks and Volumes
+
+- Disk support with virtio-block or vhost-user-block
+- Volume support with virtio-fs
+
+#### Network
+
+- Support based on virtio-net or vhost-user-net
+- Support for tap and macvtap
+
+#### Host Device Assignment
+
+- VFIO supported for passing through PCI devices
+
+#### Accessing Virtual Machines
+
+- Serial port (0x3f8) and virtio-console are supported
+- Create a PTY so that external process can later connect to it
+- Support for SSH as it directly depends on virtio-net support
+
+#### Confidential Computing
+
+- Support for SGX
+- Experimental support for TDX
+
+#### Architectures
+
+- x86_64
+- Aarch64
+
+#### Migration
+
+This should be supported eventually but it still requires some assessment of
+how to achieve it. Therefore we might not see this feature being supported for
+some time.
+
+### Unsupported features
+
+Here is the list of what will be missing compared to what libvirt/QEMU supports:
+
+#### Virtual Hardware
+
+- No CPU model emulation
+- No way to pick a type of clock
+- No way to pick a type of timer
+- No support for emulated video and graphics devices
+- No way to pick between different features like `acpi`, `apic`. We can select
+  `hyperv` though, which enables KVM Hyper-V enlightments
+- No support for emulated input device
+
+#### Disks and Volumes
+
+- No support for cdrom, floppy disk or luns
+
+#### Network
+
+- No support for emulated NICs such as e1000, e1000e, ... (which means no SLIRP)
+
+#### Accessing Virtual Machines
+
+- No support for VNC
+
+### Guest Agent
+
+Features related to the ability of running a dedicated agent in the guest have
+not been tested yet. The existing QEMU agent must be evaluated to see if it
+could work and be reused directly with Cloud Hypervisor. If that's not the case,
+an agent program would have to be developed for operations like `GuestPing`,
+`ListInterfaces`, ...
+

```

Comment:
The guest agent is also used for performing consistent snapshot while the VM is running. This is an important feature. So, until we don't have a solution for the guest agent, I'd list that we can only perform offline snapshots for cloud hypervisor

Reply:
By "offline" vs "online", do you mean "snapshot" vs "live migration"? I'm asking because Cloud Hypervisor supports both snapshot and live migration without the need for any agent.

Reply:
No, with online and offline I mean taking the snapshot while the VM is running or not. If the VM is running, it could still continue to write on the disk you are trying to snapshot. The guest agent perform the fsfreeze and this guarantee the consistency of the snapshot,

Reply:
You don't have this inconsistency if the VM isn't running, but this has the limitation that you need to shut down the VM before taking the snapshot.

Reply:
Here, we are talking about the storage snapshot performed by CSI

Reply:
Ah thank you for the clarification, that makes sense.